File: ch_07/quicksort.cpp
#include <iostream>
int partition(int [], int, int);

void quicksort(int a[], int start, int end)
{
    if (start < end)
    {
        int p = partition(a, start, end);
        quicksort(a, start, p - 1);
        quicksort(a, p + 1, end);
    }
}

int partition(int a[], int start, int end)
{   
    int pivot = a[end];
    int i = start - 1;

    for (int j = start; j < end; j++)
        if (a[j] <= pivot)
            std::swap(a[++i], a[j]);
    std::swap(a[++i], a[end]);

    return i;
}

int main()
{
    int a[] = { 1, 3, 5, 2, 4, 6, 9, 8, 0};
    quicksort(a, 0, 8);
    for (int i = 0; i < 9; i++)
        std::cout << i << " ";
    std::cout << std::endl;
}

File: ch_15/quicksort.py
def quicksort(items, start, end):
    if (start < end):
        q = partition(items, start, end)
        quicksort(items, start, q - 1)
        quicksort(items, q + 1, end)

def partition(a, start, end):
    pivot = a[end]
    i = start - 1
    for j in range(start, end):
        if (a[j] <= pivot):
            i = i + 1;
            a[j], a[i] = a[i], a[j]
    i = i + 1    
    a[i], a[end] = a[end], a[i]
    return i

if __name__ == "__main__":
    items = [1, 3, 5, 4, 2, 9, 8, 6]
    quicksort(items, 0, len(items) - 1)
    print items
